include/lsst/afw/geom/Endpoint.h

```diff
@@ -0,0 +1,303 @@
+(other forms are possible, but this is the one chosen):
+- std::vector<double> for a single point
+- ndarray<double, 2, 2> for an array of points
+
+Endpoints are designed as helper classes for Transform. Each transform has a two endpoints:
+one for input data and one for output data.
+
+Endpoint also provides two methods to work with ast::Frames:
+- normalizeFrame verifies that a frame is the correct type, and adjusts its settings if necessary
+- makeFrame creates a new frame with the correct type and settings
+
+@tparam PointT  LSST data type for one point
+@tparam ArrayT  LSST data type for an array of points
+*/
+
+template <typename PointT, typename ArrayT>
+class BaseEndpoint {
+public:
+    using Point = PointT;
+    using Array = ArrayT;
+
+    /**
+    Construct a BaseEndpoint
+
+    @param[in] nAxes  the number of axes in a point
+    */
+    explicit BaseEndpoint(int nAxes) : _nAxes(nAxes) {}
+
+    virtual ~BaseEndpoint() {};
+
+    int getNAxes() const { return _nAxes; }
+
+    /**
+    Return the number of points in an array
+    */
+    virtual int getNPoints(Array const &arr) const = 0;
+
+    /**
+    Get raw data from a single point
+
+    @param[in] point  data for a single point
+    @return the values in the point as a vector of size NAxess
+
+    @throw lsst::pex::exceptions::InvalidParameterError if the point has the wrong number of axes
+    */
+    virtual std::vector<double> dataFromPoint(Point const & point) const = 0;
+
+    /**
+    Get raw data from an array of points
+    
+    @param[in] arr  Array of points
+    @return the data as a 2-D ndarray array [nPoints, nAxes] in C order,
+        so the in-memory view is, for example, x0, y0, x1, y1, x2, y2, ...
+
+    @throw lsst::pex::exceptions::InvalidParameterError if the array has the wrong nAxes dimension
+    */
+    virtual ndarray::Array<double, 2, 2> dataFromArray(Array const & arr) const = 0;
+
+    /**
+    Get a single point from raw data
+
+    @param[in] data  Data as a vector of length NAxes
+    @return the corresponding point
+    */
+    virtual Point pointFromData(std::vector<double> const & data) const = 0;
+
+    /**
+    Get an array of points from raw data
+    
+    @param[in] data  Raw data for an array of points, as a 2-D ndarray array [nPoints, nAxes] in C order,
+        so the in-memory view is, for example, x0, y0, x1, y1, x2, y2, ...
+    @return an array of points
+
+    @throw lsst::pex::exceptions::InvalidParameterError if the array has the wrong nAxes dimension
+    */
+    virtual Array arrayFromData(ndarray::Array<double, 2, 2> const & data) const = 0;
+
+    /**
+    Create a Frame that can be used with this end point in a Transform
+    */
+    virtual std::shared_ptr<ast::Frame> makeFrame() const;
+
+    /**
+    Adjust and check the frame as needed.
+
+    Do not obother to check the number of axes because that is done elsewhere.
+
+    The base implementation does nothing.
```

Comment:
This is a default behavior that derived classes inherit and can choose to override. As such it might do something useful, but in this case it does nothing.